source/extensions/transport_sockets/tls/ocsp/ocsp.cc

```diff
@@ -0,0 +1,307 @@
+    throw EnvoyException("Failed to parse ASN.1 element tag");
+  }
+  return tag;
+}
+
+std::unique_ptr<OcspResponse> readDerEncodedOcspResponse(const std::string& der) {
+  CBS cbs;
+  CBS_init(&cbs, reinterpret_cast<const uint8_t*>(der.c_str()), der.size());
+
+  auto resp = Asn1OcspUtility::parseOcspResponse(cbs);
+  if (CBS_len(&cbs) != 0) {
+    throw EnvoyException("Data contained more than a single OCSP response");
+  }
+
+  return resp;
+}
+
+void skipResponderId(CBS& cbs) {
+  // ResponderID ::= CHOICE {
+  //    byName               [1] Name,
+  //    byKey                [2] KeyHash
+  // }
+  //
+  // KeyHash ::= OCTET STRING -- SHA-1 hash of responder's public key
+  //    (excluding the tag and length fields)
+
+  if (Asn1Utility::isOptionalPresent(cbs, nullptr, CONS | CONT | 1) ||
+      Asn1Utility::isOptionalPresent(cbs, nullptr, CONS | CONT | 2)) {
+    return;
+  }
+
+  throw EnvoyException(absl::StrCat("Unknown choice for Responder ID: ", parseTag(cbs)));
+}
+
+} // namespace
+
+OcspResponse::OcspResponse(OcspResponseStatus status, ResponsePtr&& response)
+    : status_(status), response_(std::move(response)) {}
+
+BasicOcspResponse::BasicOcspResponse(ResponseData data, std::string signature_alg,
+                                     std::vector<uint8_t> signature)
+    : data_(data), signature_alg_(signature_alg), signature_(std::move(signature)) {}
+
+const std::string BasicOcspResponse::OID = "1.3.6.1.5.5.7.48.1.1";
+
+ResponseData::ResponseData(Envoy::SystemTime produced_at,
+                           std::vector<SingleResponse> single_responses)
+    : produced_at_(produced_at), single_responses_(std::move(single_responses)) {}
+
+SingleResponse::SingleResponse(CertId cert_id, CertStatus status, Envoy::SystemTime this_update,
+                               absl::optional<Envoy::SystemTime> next_update)
+    : cert_id_(cert_id), status_(status), this_update_(this_update), next_update_(next_update) {}
+
+CertId::CertId(std::string serial_number, std::string alg_oid, std::string issuer_name_hash,
+               std::string issuer_public_key_hash)
+    : serial_number_(serial_number), alg_oid_(alg_oid), issuer_name_hash_(issuer_name_hash),
+      issuer_public_key_hash_(issuer_public_key_hash) {}
+
+OcspResponseWrapper::OcspResponseWrapper(std::string der_response, TimeSource& time_source)
+    : raw_bytes_(std::move(der_response)), response_(readDerEncodedOcspResponse(raw_bytes_)),
+      time_source_(time_source) {
+
+  if (response_->response_ == nullptr) {
+    throw EnvoyException("OCSP response has no body");
+  }
+
+  // We only permit a 1:1 of certificate to response
+  if (response_->response_->getNumCerts() != 1) {
+    throw EnvoyException("OCSP Response must be for one certificate only");
+  }
+
+  auto& this_update = response_->response_->getThisUpdate();
+  if (time_source_.systemTime() < this_update) {
+    DateFormatter formatter("%E4Y%m%d%H%M%S");
```

Comment:
This date format string is used in several places. Find a common home for it in a header?